cpp/src/jni/dataset/concurrent_map.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ */
+
+#ifndef JNI_ID_TO_MODULE_MAP_H
+#define JNI_ID_TO_MODULE_MAP_H
+
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+
+#include "jni.h"
+#include "arrow/util/macros.h"
+
+namespace arrow {
```

Comment:
Please do not copy and paste this file, factor it out to a common place that can be used by all JNI implementations.

Reply:
Maybe move concurrent_map.h out into `src/jni/`? By the way, should we somehow migrate `src/gandiva/jni` to `src/jni/gandiva`, or migrate `src/jni` to `arrow/jni`? It seems that currently they are organized inconsistently.

Reply:
see comments below, I don't think we should be using concurrent_map in general, especially for new use cases so we can drop it for now.

> By the way, should we somehow migrate src/gandiva/jni to src/jni/gandiva, or migrate src/jni to src/arrow/jni? It seems that currently they are organized inconsistently.
I don't have a strong opinion here probably best to discuss this on the mailing list.

Reply:
the file has been removed